clippy_lints/src/methods/map_flatten.rs

```diff
@@ -68,16 +71,44 @@ pub(super) fn check<'tcx>(
         },
     };
 
-    let func_snippet = snippet(cx, map_arg.span, "..");
-    let hint = format!(".and_then({})", func_snippet);
+    let mut applicability = Applicability::MachineApplicable;
+    let sugg_span = expr.span.with_lo(map_span.lo());
     let lint_info = format!("called `map(..).flatten()` on an `{}`", caller_type);
-    span_lint_and_sugg(
-        cx,
-        MAP_FLATTEN,
-        expr.span.with_lo(recv.span.hi()),
-        &lint_info,
-        "try using `and_then` instead",
-        hint,
-        Applicability::MachineApplicable,
-    );
+    let help_msgs = [
+        "try replacing `map` with `and_then` instead",
+        "and remove the final `.flatten()`",
```

Comment:
I think here we should remove the *final* to not be confusing for longer changes that have multiple `flatten()` and other magic. (I know that I suggested a message like this, but that came to mind afterwards :) )